### PR TITLE
Define spectrum processing functions

### PR DESCRIPTION
**Pull request by florian-huber**
Add necessary filtering functions (or restructure the ones we already had). Aim is to address #60.

- minimal processing is now done via `spectrum_processing_minimal()`.
- further processing to prepare for Spec2Vec calculation (but should also work for all other scores): `spectrum_processing_s2v()`

Bit unclear, but this also contains a refactored CI workflow including sonarcloud which I just added to the main branch.

### CODE REVIEW

File: ms2query/spectrum_processing.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        Set lower threshold for peak intensity. Default is 0.001.
+    max_mz_required
+        Only peaks <= max_mz_required will be counted to check if spectrum
+        contains sufficient peaks to be considered.
+    """
+    settings = set_minimal_processing_defaults(**settings)
+    spectrum = normalize_intensities(spectrum)
+    spectrum = select_by_intensity(spectrum, intensity_from=settings["intensity_from"])
+    spectrum = select_by_mz(spectrum, mz_from=settings["mz_from"], mz_to=np.inf)
+    spectrum = require_peaks_below_mz(spectrum, n_required=settings["n_required_below_1000"],
```
Comment (niekdejonge) [inline ms2query/spectrum_processing.py:61]: Looks good! Just to be sure I understood this correctly, this should be run before loading the spectra into sqlite right? Since none of these processing steps are negative for any other function like ms2ds or cos scores.

File: ms2query/spectrum_processing.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+def spectrum_processing_s2v(spectrum: SpectrumType,
+                            **settings: Dict[str, Union[int, bool]]) \
+    -> Union[SpectrumType, None]:
+    """Spectrum processing required for computing Spec2Vec scores.
+
+    Args:
+    ----------
+    spectrum:
+        Spectrum to process
+    mz_from
```
Comment (niekdejonge) [inline ms2query/spectrum_processing.py:102]: A lot of these variables are not used anymore, so they can be removed.

File: .github/workflows/CI_build.yml [hunk 3]
```diff
@@ -7,6 +7,36 @@ on:
 
 jobs:
 
+  thorough_check:
```
Comment (niekdejonge) [inline .github/workflows/CI_build.yml:10]: I did not look into the workflows, since I do not have enough experience with this to say anything useful about these changes.

File: ms2query/spectrum_processing.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+def set_spec2vec_defaults(**settings: Dict[str, Union[int, bool]]) \
+        -> Dict[str, Union[int, bool]]:
+    """Set spec2vec default argument values"(where no user input is given)".
+
+    Args
+    ----------
+    **settings:
+        Change default settings
+    """
+    defaults = {"mz_from": 0,
```
Comment (niekdejonge) [inline ms2query/spectrum_processing.py:74]: Not all defaults are needed , so both in spectrum_processing_spec2vec and here, these variables can be removed

File: tests/test_spectrum_processing.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    intensities = np.array([0, 1, 10, 100], dtype="float")
+    spectrum_in = Spectrum(mz=mz, intensities=intensities)
+
+    spectrum = require_peaks_below_mz(spectrum_in, n_required=4, max_mz=400)
+
+    assert spectrum is None, \
+        "Expected None since peaks with mz<=max_mz is less than n_required."
+
+
+def test_require_peaks_below_empty_spectrum():
```
Comment (niekdejonge) [inline tests/test_spectrum_processing.py:43]: Nice thorough testing of test_require_peaks_below_mz

File: tests/test_spectrum_processing.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+def test_spectrum_processing_minimal_default():
+    spectrum_in = Spectrum(mz=np.array([110, 220, 330, 440], dtype="float"),
+                           intensities=np.array([10, 1, 10, 100], dtype="float"))
+    spectrum = spectrum_processing_minimal(spectrum_in)
+
+    assert spectrum is None, \
+        "Expected None because the number of peaks (4) is less than the default threshold (10)."
+
+
+def test_spectrum_processing_minimal_set_n_required():
```
Comment (niekdejonge) [inline tests/test_spectrum_processing.py:59]: Maybe rename function to test_spectrum_processing_minimal_normalization_required

File: tests/test_spectrum_processing.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+
+def test_require_peaks_below_empty_spectrum():
+    spectrum_in = None
+    spectrum = require_peaks_below_mz(spectrum_in)
+
+    assert spectrum is None, "Expected different handling of None spectrum."
+
+
+def test_spectrum_processing_minimal_default():
```
Comment (niekdejonge) [inline tests/test_spectrum_processing.py:50]: I am not really sure what the correct way of testing a function that only calls other functions is, since all functions used already have their own test functions. One way I could think of is running one test for each function used (but not all tests for each function), to make sure they are all called. In that case a few test functions would be missing, but the current solution of testing one positive and one negative case might already be good enough.

File: tests/test_spectrum_processing.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+# TODO: uncomment once matchms n_max issue is fixed (#177 in matchms)
+# def test_spectrum_processing_s2v_set_n_max():
+#     """Test processing an individual spectrum with spectrum_processing_s2v"""
+#     spectrum_in = Spectrum(mz=np.array([5, 110, 220, 330, 440], dtype="float"),
+#                            intensities=np.array([0.1, 0.2, 0.1, 1, 0.5], dtype="float"),
+#                            metadata={"parent_mass": 250.0,
+#                                      "precursor_mz": 240.0})
+#     spectrum = spectrum_processing_s2v(spectrum_in, n_max=4)
+#     assert isinstance(spectrum, Spectrum), "Expected output to be Spectrum."
+#     assert spectrum.peaks == spectrum_in.peaks[1:], "Expected spectrum peaks to be unchanged by function"
```
Comment (niekdejonge) [inline tests/test_spectrum_processing.py:99]: Change the comment to something like "Expected only the first peak to be removed"

File: ms2query/spectrum_processing.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    allowed_missing_percentage
+        Allowed percentage of the entire spectrum that is allowed to not be
+        covered my the pretrained Spec2Vec model.
+    """
+    settings = set_spec2vec_defaults(**settings)
+    spectrum = reduce_to_number_of_peaks(spectrum,
+                                         n_required=settings["n_required"],
+                                         ratio_desired=settings["ratio_desired"],
+                                         n_max=settings["n_max"])
+    if spectrum is None:
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply ms2query/spectrum_processing.py:130]: True, good point. It's not needed for this function, so I took it out (I didn't add a assert though, because it would anyway then fail with None).

File: ms2query/spectrum_processing.py (hunk 2 above)
Comment (florian-huber) [reply ms2query/spectrum_processing.py:102]: Thanks. Removed the old ones.

File: .github/workflows/CI_build.yml (hunk 3 above)
Comment (florian-huber) [reply .github/workflows/CI_build.yml:10]: That's fine. It should have been separate anyway, but I mostly copied it from another repo where it had been reviewed.

File: ms2query/spectrum_processing.py (hunk 1 above)
Comment (florian-huber) [reply ms2query/spectrum_processing.py:61]: Yes, that's correct. This should become the minimal processing required for all data going into the sqlite.

File: tests/test_spectrum_processing.py (hunk 5 above)
Comment (florian-huber) [reply tests/test_spectrum_processing.py:43]: Thanks.

File: ms2query/spectrum_processing.py (hunk 4 above)
Comment (florian-huber) [reply ms2query/spectrum_processing.py:74]: Yes that's true. Here it was a bit on purpose to keep everything spec2vec related here. But maybe that's confusing, so I removed most things now and we can add them later. I kept a tiny bit of redundancy and left 2 overlapping settings with the basic processing step to have the entire set of processing defaults included.

File: tests/test_spectrum_processing.py (hunk 6 above)
Comment (florian-huber) [reply tests/test_spectrum_processing.py:59]: Good catch. Changed it.

File: tests/test_spectrum_processing.py (hunk 7 above)
Comment (florian-huber) [reply tests/test_spectrum_processing.py:50]: That is precisely why I added `sonarcloud`. I gives a very nice overview of the test coverage of our code. If you click on the code coverage link there it will highlight all uncovered lines. That also includes if a function is run from another function, so it is not strictly necessary to test every function individually. I usually do that when I want to test something very specific or aim for coverage of some conditions which are otherwise missing.

Comment (florian-huber) [reply tests/test_spectrum_processing.py:50]: Oh, and would often also just decide pragmatically. One test too much (e.g. already covered elsewhere) is not too bad as long as this doesn't become excessive. And sonarcloud will then point us to missing aspects that are not tested yet.

File: tests/test_spectrum_processing.py (hunk 8 above)
Comment (florian-huber) [reply tests/test_spectrum_processing.py:99]: True, good catch. I changed it.

File: ms2query/spectrum_processing.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (niekdejonge) [reply ms2query/spectrum_processing.py]: I think it would not fail anyway, since the reduce_to_number_of_peaks and add_losses are able to handle None as an input spectrum and it will return None again. See https://github.com/matchms/matchms/blob/master/matchms/filtering/reduce_to_number_of_peaks.py
So maybe good to still add the assert function.
